presto-main/src/main/java/com/facebook/presto/operator/OperationTimer.java

```diff
@@ -0,0 +1,285 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.google.common.base.Ticker;
+import io.airlift.units.Duration;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public class OperationTimer
+{
+    private static final ThreadMXBean THREAD_MX_BEAN = ManagementFactory.getThreadMXBean();
+    private static final long DEFAULT_MIN_NANOS_FOR_CPU_TIMING = new Duration(250, TimeUnit.MILLISECONDS).roundTo(NANOSECONDS);
+
+    private final Ticker ticker;
+    private final long minNanosForCpuTiming;
+
+    private final long wallStart;
+    private final long cpuStart;
+    private final long userStart;
+
+    private long intervalWallStart;
+
+    private long intervalUnaccountedCpuWallTime;
+    private long intervalCpuStart;
+    private long intervalUserStart;
+
+    private OperationTiming overallTiming;
+    private final Map<OperationTiming, InternalTiming> timings = new HashMap<>();
+
+    public OperationTimer()
+    {
+        this(Ticker.systemTicker(), DEFAULT_MIN_NANOS_FOR_CPU_TIMING);
+    }
+
+    public OperationTimer(Ticker ticker, long minNanosForCpuTiming)
+    {
+        this.ticker = requireNonNull(ticker, "ticker is null");
+        this.minNanosForCpuTiming = minNanosForCpuTiming;
+
+        wallStart = ticker.read();
+        cpuStart = currentThreadCpuTime();
+        userStart = currentThreadUserTime();
+
+        intervalWallStart = wallStart;
+        intervalCpuStart = cpuStart;
+        intervalUserStart = userStart;
+    }
+
+    public void setOverallTimer(OperationTiming overallTiming)
+    {
+        requireNonNull(overallTiming, "overallTiming is null");
+        checkState(this.overallTiming == null, "Overall timer already set");
+        this.overallTiming = overallTiming;
+    }
+
+    public void recordOperationComplete(OperationTiming timing)
+    {
+        requireNonNull(timing, "timing is null");
+        InternalTiming internalTiming = timings.computeIfAbsent(timing, InternalTiming::new);
```

Comment:
This one seems to be brittle. How about having a `Map<OperatorId, InternalTiming>`, and having `recordOperationComplete` accept `int operatorId`

Reply:
Each operator has multiple timing objects.  The idea is they create them and use each one for the timings they want.  The `OperationTiming` class in nested in this class, so we can rely on the behavior of equals an hash code.